### PR TITLE
Enhance backward process

### LINKED ISSUE #15283: 用户使用momentum optimizer时出现cast op输入不存在错误
用户反馈在使用momentum optimizer时append backward出现cast op错误，用其他optimizer未出现错误。

Paddle版本release 1.2。

code:
```python
def balanced_cross_entropy_bootstrap_loss(prediction, gt_org, average_flag = True, bootstrapped_threshold=0.95):
    class_num = prediction.shape[1]
    #get gt_mask
    gt = fluid.layers.reshape(gt_org, [0, 1, gt_org.shape[-2], gt_org.shape[-1]])

    gt_mask = (gt==0).astype('float32')
    for i in range(1, class_num):
        gt_mask_i = (gt==i).astype('float32')
        gt_mask = fluid.layers.concat([gt_mask, gt_mask_i], axis=1)

    #get valid_mask
    valid_mask =  (gt < class_num).astype('float32')
    valid_mask = fluid.layers.expand(valid_mask, [1, class_num, 1, 1])
    #bootstrap
    prediction = prediction.astype('float32')
    pred = fluid.layers.argmax(prediction, axis=1).astype('float32')
    pred = fluid.layers.reshape(pred, [0, 1, pred.shape[-2], pred.shape[-1]])
    bootstrap_valid1 = (pred == gt).astype('float32')
    bootstrap_valid1 = fluid.layers.expand(bootstrap_valid1, [1, class_num, 1, 1])
    bootstrap_valid2 = (prediction > bootstrapped_threshold).astype('float32')
    bootstrap_valid = fluid.layers.elementwise_mul(bootstrap_valid2, bootstrap_valid1)
    bootstrap_mask = (1.0 - bootstrap_valid).astype('float32')
    #print(bootstrap_mask)
    #print(valid_mask)
    valid_mask = fluid.layers.elementwise_mul(bootstrap_mask, valid_mask).astype('float32')

    #get valid gt_mask
    #gt_mask = fluid.layers.elementwise_mul(gt_mask, valid_mask).astype('float32')

    #get balanced weight
    count_num = fluid.layers.reduce_sum(gt_mask, dim=[2, 3]).astype('float32')
    count_num = fluid.layers.elementwise_max(count_num, fluid.layers.assign(np.array([0.00001], dtype=np.float32)))
    balanced_weight = 1 - count_num/fluid.layers.reduce_sum(count_num, dim=1).astype('float32') #count_num [n,c]
    balanced_gt_mask = fluid.layers.elementwise_mul(gt_mask, balanced_weight, axis=0).astype('float32') #gt_mask=[n,c,h,w] balanced_weight=[n,c]

    prediction = fluid.layers.softmax(prediction).astype('float32')
    prediction = fluid.layers.elementwise_max(prediction, fluid.layers.assign(np.array([0.00001], dtype=np.float32)))
    prediction = fluid.layers.log(prediction).astype('float32')
    loss_mask = fluid.layers.elementwise_mul(prediction, balanced_gt_mask, axis=0)

    if average_flag:
        valid_loss = fluid.layers.reduce_mean(fluid.layers.elementwise_mul(loss_mask, valid_mask))
        #valid_loss = fluid.layers.reduce_mean(loss_mask)
    else:
        valid_loss = fluid.layers.reduce_sum(fluid.layers.elementwise_mul(loss_mask, valid_mask, axis=0))
    return    valid_mask,  -1*valid_loss
```

错误log（编译期错误）：
```
The input of cast op must be set.
```

解决方案：将某些不需要梯度的变量设置为stop_gradient=True，例如
```
bootstrap_mask.stop_gradient = True
valid_mask.stop_gradient = True
```

### PR DESCRIPTION
**Pull request by chengduoZH**
Prune the unnecessary op in backward.
Related PR https://github.com/PaddlePaddle/Paddle/pull/15955

Fix https://github.com/PaddlePaddle/Paddle/issues/15283

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_backward.py [hunk 1]
```diff
@@ -30,6 +30,17 @@ def simple_net1():
     return loss
 
 
+def simple_net2():
+    x = fluid.layers.data(name='image', shape=[784], dtype='float32')
+    label = fluid.layers.data(name='label', shape=[1], dtype='int64')
+    feature = fluid.layers.fc(input=x, size=10, act=None)
+    label = fluid.layers.cast(label, dtype="float32")
+    label = fluid.layers.cast(label, dtype='int64')
+    loss = fluid.layers.cross_entropy(input=feature, label=label)
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_backward.py:40]: Note that the label is not persistable in `fluid.layers.cross_entropy`.

Comment (luotao1) [reply python/paddle/fluid/tests/unittests/test_backward.py:40]: 这个解释https://github.com/PaddlePaddle/Paddle/pull/18700/files#r305269149 可以加入单测代码中么？
这个单测的意思是，如果有并不需要的label，也能保持正确么？

File: python/paddle/fluid/backward.py [hunk 2]
```diff
@@ -247,6 +247,104 @@ def _op_can_be_removed_(op_desc, no_grad_set):
     return op_descs
 
 
+def __find_not_need_ops(grad_op_descs, ops, input_grad_names_set):
```
Comment (luotao1) [inline python/paddle/fluid/backward.py]: 1. 可以多加一些注释，说一下整体剪枝的逻辑么，比如是先转成图再剪？
2. 会有性能上的影响么？

Comment (Xreki) [reply python/paddle/fluid/backward.py]: 同@luotao02 
虽然issue里面有详情，也建议在PR里面描述一下。

Comment (chengduoZH) [reply python/paddle/fluid/backward.py]: > 可以多加一些注释，说一下整体剪枝的逻辑么，比如是先转成图再剪？

Done

> 会有性能上的影响么？

时间复杂度是O(N)，与之前构建backward过程的时间复杂度是一样的。

File: python/paddle/fluid/backward.py [hunk 3]
```diff
@@ -247,6 +247,116 @@ def _op_can_be_removed_(op_desc, no_grad_set):
...
+    """
+    Pruning Program with Structural Analysis Method of Computational Graph.
+    The nodes of the computational graph composed of backward OPS should be
+    interconnected. If there are unconnected sub-graphs in the computational graph,
+    these sub-graphs should be cut off.
```
Comment (qingqing01) [inline python/paddle/fluid/backward.py:255]: Maybe better to add comments for grad_op_descs, ops, input_grad_names_set

Comment (chengduoZH) [reply python/paddle/fluid/backward.py:255]: Done

**Review by qingqing01 — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM for update `backward.py`
